src/golibmc.go

```diff
@@ -184,61 +310,248 @@ func New(servers []string, noreply bool, prefix string, hashFunc int, failover b
 	}
 
 	failoverInt := 0
-	if failover {
+	if client.failover {
 		failoverInt = 1
 	}
 
 	C.client_init(
-		client._imp,
+		cn._imp,
 		(**C.char)(unsafe.Pointer(&cHosts[0])),
 		(*C.uint32_t)(unsafe.Pointer(&cPorts[0])),
 		C.size_t(n),
 		(**C.char)(unsafe.Pointer(&cAliases[0])),
 		C.int(failoverInt),
 	)
 
-	client.configHashFunction(int(hashFunctionMapping[hashFunc]))
-	client.servers = servers
-	client.prefix = prefix
-	client.noreply = noreply
-	client.disableLock = disableLock
-	return
+	cn.configHashFunction(int(hashFunctionMapping[client.hashFunc]))
+	if client.retryTimeout > 0 {
+		C.client_config(cn._imp, RetryTimeout, client.retryTimeout)
+	}
+	if client.pollTimeout > 0 {
+		C.client_config(cn._imp, PollTimeout, client.pollTimeout)
+	}
+	if client.connectTimeout > 0 {
+		C.client_config(cn._imp, ConnectTimeout, client.connectTimeout)
+	}
+	return &cn, nil
 }
 
-// SimpleNew to create a memcached client with default params
-func SimpleNew(servers []string) (client *Client) {
-	noreply := false
-	prefix := ""
-	hashFunc := HashCRC32
-	failover := false
-	disableLock := false
-	return New(
-		servers, noreply, prefix, hashFunc,
-		failover, disableLock,
-	)
+func (client *Client) putConn(cn *conn) error {
+	client.lk.Lock()
+	if cn.badConn {
+		client.lk.Unlock()
+		err := cn.quit()
+		if err != nil {
+			log.Println("Faild cn.close", err)
+		}
+		return err
+	}
+	client.putConnLocked(cn)
+	client.lk.Unlock()
+	return nil
 }
 
-func finalizer(client *Client) {
-	C.client_destroy(client._imp)
+func (client *Client) putConnLocked(cn *conn) bool {
+	if client.closed {
+		return false
+	}
+	if client.maxOpen > 0 && client.maxOpen < client.numOpen {
+		return false
+	}
+	if len(client.connRequests) > 0 {
+		var req chan *conn
+		var reqKey uint64
+		for reqKey, req = range client.connRequests {
+			break
+		}
+		delete(client.connRequests, reqKey)
+		req <- cn
+	} else {
+		client.freeConns = append(client.freeConns, cn)
+		client.startCleanerLocked()
+	}
+	return true
 }
 
-func (client *Client) lock() {
-	if !client.disableLock {
+func (client *Client) conn(ctx context.Context) (*conn, error) {
+	cn, err := client._conn(ctx, true)
+	if err == nil {
+		return cn, nil
+	}
+	if err == ErrBadConn {
+		return client._conn(ctx, false)
+	}
+	return cn, err
+}
+
+func (client *Client) _conn(ctx context.Context, useFreeClient bool) (*conn, error) {
+	client.lk.Lock()
+	if client.closed {
+		client.lk.Unlock()
+		return nil, ErrMemcachedClosed
+	}
+	// Check if the context is expired.
+	select {
+	default:
+	case <-ctx.Done():
+		client.lk.Unlock()
+		return nil, ctx.Err()
+	}
+	lifetime := client.maxLifetime
+
+	var cn *conn
+	if useFreeClient && len(client.freeConns) > 0 {
+		cn, client.freeConns = client.freeConns[0], client.freeConns[1:]
```

Comment:
this strategy will let the slice's under storage increase unnecessarily？

ref: https://blog.golang.org/go-slices-usage-and-internals 'A possible "gotcha"'
prefer https://github.com/golang/go/blob/master/src/database/sql/sql.go#L1023

Reply:
I see. I learned a lot. Fixed.
https://github.com/douban/libmc/pull/57/commits/55b67a90964e5a97f570ed8807ba9d7caedafb2c

Reply:
you're welcome